### PR TITLE
mgr/dashboard: Multi cluster improvements

### PR DESCRIPTION
**Pull request by aaSharma14**
1. Since, after adding a new cluster we add its IP to the list of prometheus federation targets. The metrics of the added cluster take some time to be received at dashboard's end hence added a loading message for few seconds to show the correct data.
2. On disconnecting a cluster, show a message to the user that the disconnected cluster's data will be removed in around 5 minutes.
4. On disconnecting the only cluster connected, the context switcher was not getting removed, fixed that as well
5. Allow the user to enter the prometheus api url in the multi-cluster form in case of token authentication. This should fix adding cluster with a token.
6. Hide the Verify Connection button when the connection is verified.
7. Reconnect cluster was not working because of a typo..Fixed that as well.
8. The Cluster Capacity Card was not showing correct data ..fixed in this PR.
9. The graph timer selection was not working properly..fixed in this PR
10. Non-existent user login attempts are being added to the Multi-Cluster list - fixed in this PR
11. Charts were not updating live and top 5 values were not sorted - fixed in this PR.
12. Do not allow the user to do multi cluster management of remote cluster from the local cluster
13. Dashboard becomes unresponsive if the token of hub cluster expires..fixed in this PR.
14. Fix charts size in the Overview Page

Fixes: https://tracker.ceph.com/issues/64880

[Screencast from 2024-03-06 13-25-06.webm](https://github.com/ceph/ceph/assets/66050535/42cceb41-329a-4611-8a0a-80f8977118d7)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/cephadm/cephadmlib/daemons/monitoring.py [hunk 1]
```diff
@@ -45,6 +45,7 @@ class Monitoring(ContainerDaemonForm):
             'args': [
                 '--config.file=/etc/prometheus/prometheus.yml',
                 '--storage.tsdb.path=/prometheus',
+                '--query.lookback-delta=1m',
```
Comment (nizamial09) [inline src/cephadm/cephadmlib/daemons/monitoring.py:48]: we should really look at the performance impact of this change. with ceph usually people will have huge number of metrics coming in at regular intervals and if the lookback-delta is being reduced then as i understand it, this can put a load on the targets (by frequently trying to pull newer metrics) and that can cause issues with the performance. Also I seen stuffs like the query latency might also get increased with this. It might not come up in our local testing with docker based environment but it can have potential risks added to it in a real environment. Have you done any assessment on how impactful this would be for us?

Comment (cloudbehl) [reply src/cephadm/cephadmlib/daemons/monitoring.py:48]: @aaSharma14 Lets use default value and not change it.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html [hunk 2]
```diff
@@ -50,6 +51,15 @@ <h4 class="mt-3">Upgrade your current cluster to a multi-cluster setup effortles
...
      *ngIf="isMultiCluster; else emptyCluster">
   <ng-container *ngIf="!loading; else loadingTpl">
+    <cd-alert-panel type="info"
+                    spacingClass="mb-3"
+                    [showTitle]="false"
+                    size="slim"
+                    *ngIf="showMessage"
+                    (dismissed)="showMessage = false"
+                    [dismissible]="true">
+      <p>Please note that the data for the disconnected cluster will remain accessible for a duration of around 5 minutes. After this period, it will be automatically removed.</p>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html:61]: let's be precise here with the duration.
```suggestion
      <p>Please note that the data for the disconnected cluster will be visible for a duration of 5 minutes. After this period, it will be automatically removed.</p>
```

Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html:61]: btw, this message will show up on every refresh if I am not wrong? or does it go away completely after you dismiss it. We want the latter. Can we configure it like that?

**Review by nizamial09 — COMMENTED** [review]
there is an empty `cephadm.py` file here which should be removed from this change.

**Comment by nizamial09** [comment]
Can you also create tracker and attach it here? Thanks

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts [hunk 3]
```diff
@@ -170,6 +175,24 @@ export class MultiClusterComponent implements OnInit {
...
+          this.router.navigateByUrl('/pool', { skipLocationChange: true }).then(() => {
+            this.router.navigate([currentRoute]);
+          });
+        } else {
+          this.router.navigateByUrl('/', { skipLocationChange: true }).then(() => {
+            this.router.navigate([currentRoute]);
+          });
+        }
+        this.getPrometheusData(this.prometheusService.lastHourDateObject);
+      }, 20000)
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts:194]: The interval here could be set with a constant value, for better clarity

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html [hunk 4]
```diff
@@ -10,6 +10,12 @@
           [formGroup]="deletionForm"
           novalidate>
       <div class="modal-body">
+        <cd-alert-panel *ngIf="showMultiClusterMessage"
+                        type="info"
+                        spacingClass="mb-3"
+                        i18n>
+          <p>Please note that the data for the disconnected cluster will remain accessible for a duration of around 5 minutes. After this period, it will be automatically removed.</p>
+        </cd-alert-panel>
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html:18]: Since this is a shared component, lets keep it generic
```suggestion
        <cd-alert-panel *ngIf="showInfoMessage"
                        type="info"
                        spacingClass="mb-3"
                        i18n>
          <p>{{infoMessage}}</p>
        </cd-alert-panel>
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.ts [hunk 5]
```diff
@@ -25,6 +25,8 @@ export class CriticalConfirmationModalComponent implements OnInit {
   itemDescription: 'entry';
   itemNames: string[];
   actionDescription = 'delete';
+  showMultiClusterMessage = false;
+
```
Comment (Pegonzal) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.ts:29]:
```suggestion
  showInfoMessage = false;
  infoMessage: string
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html (hunk 2 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html:61]: I don't really get the need to show this message here, is not enought to have it in the modal for cluster discconection?

Also, don't forget to add internationalization (i18n)

**Review by Pegonzal — COMMENTED** [review]
Some comments from my side:

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html [hunk 6]
```diff
@@ -42,6 +42,7 @@ <h4 class="mt-3">Upgrade your current cluster to a multi-cluster setup effortles
         <i class="mx-auto d-block"
            [ngClass]="[icons.large3x, icons.spinner, icons.spin]">
         </i>
+        <p class="text-center mt-3">Loading data, Please wait...</p>
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html:45]: Hey @Pegonzal , I added this because it was just a lone spinner and didn't had any text. I wanted to show here that the Data is being loaded.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts (hunk 3 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts]: Done. Thanks @Pegonzal

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html (hunk 4 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html]: Thanks, updated

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html (hunk 2 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html]: Thanks @nizamial09 , Updated the PR. @Pegonzal , we want to show this message here so that the user doesn't get confused seeing wrong data, so just to double check the behavior this was added.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html:45]: i18n should  be applied then

[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.html]: Updated. Thanks

**Comment by cloudbehl** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts [hunk 7]
```diff
@@ -223,13 +258,35 @@ export class PrometheusService {
...
+                  ].includes(queryName)
+                ) {
+                  const request = this.getMultiClusterQueryRangeData({
+                    params: encodeURIComponent(query),
+                    start: clusterUtilizationChartSelectedTime['start'],
+                    end: clusterUtilizationChartSelectedTime['end'],
+                    step: clusterUtilizationChartSelectedTime['step']
+                  });
+                  requests.push(request);
+                }
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:289]: **THIS IS SERVICE AND IT SHOULD NOT HAVE COMPONENT CODE!**

Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts]: Updated the PR

**Comment by aaSharma14** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html [hunk 8]
```diff
@@ -110,6 +118,33 @@
...
+                   name="prometheusApiUrl"
+                   class="form-control"
+                   type="text"
+                   formControlName="prometheusApiUrl">
+            <span class="invalid-feedback"
+                  *ngIf="remoteClusterForm.showError('prometheusApiUrl', frm, 'required')"
+                  i18n>This field is required.
+            </span>
+          </div>
+        </div>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html:147]: btw, we are going to disable the auth_with_token toggle later on right?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts [hunk 9]
```diff
@@ -170,17 +204,92 @@ export class MultiClusterComponent implements OnInit {
...
+        const currentRoute = this.router.url.split('?')[0];
+        if (currentRoute.includes('dashboard')) {
+          this.router.navigateByUrl('/pool', { skipLocationChange: true }).then(() => {
+            this.router.navigate([currentRoute]);
+          });
+        } else {
+          this.router.navigateByUrl('/', { skipLocationChange: true }).then(() => {
+            this.router.navigate([currentRoute]);
+          });
+        }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts:221]: i saw this particular code gets called multiple times in different components as well. So shouldn't we generalize it?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts [hunk 10]
```diff
@@ -170,17 +204,92 @@ export class MultiClusterComponent implements OnInit {
...
+        this.getPrometheusData(this.prometheusService.lastHourDateObject);
+      }, this.PROMETHEUS_DELAY);
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts:223]: why a timeout?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-area-chart/dashboard-area-chart.component.html [hunk 11]
```diff
@@ -24,7 +24,7 @@
   </div>
 
   <div class="col-9 d-flex flex-column">
-    <div class="chart mt-3">
+    <div [ngClass]="{'chart mt-3': !isMultiCluster, 'mt-3': isMultiCluster}">
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-area-chart/dashboard-area-chart.component.html:27]: why this change? if it is to fix the pie chart responsiveness?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-pie/dashboard-pie.component.ts [hunk 12]
```diff
@@ -157,8 +157,10 @@ export class DashboardPieComponent implements OnChanges, OnInit {
   private prepareRawUsage(chart: Record<string, any>, data: Record<string, any>) {
     const nearFullRatioPercent = this.lowThreshold * 100;
     const fullRatioPercent = this.highThreshold * 100;
-    const percentAvailable = this.calcPercentage(data.max - data.current, data.max);
-    const percentUsed = this.calcPercentage(data.current, data.max);
+    const max = typeof data.max === 'string' ? parseFloat(data.max) : data.max;
+    const current = typeof data.current === 'string' ? parseFloat(data.current) : data.current;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-pie/dashboard-pie.component.ts:161]: why would this come as `string` in the first place?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts [hunk 13]
```diff
@@ -169,4 +175,18 @@ export class MultiClusterService {
 
     return this.http.get<object>('api/multi-cluster/check_token_status', { params });
   }
+
+  showPrometheusDelayMessage(showDeletionMessage?: boolean) {
+    if (showDeletionMessage !== undefined) {
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts:180]: if you give the `showDeletionMessage` true or false initially, then you don't need to check whether it defined or undefined. And I don't really understand the purpose of this function. You are just returning whatever value which is passed to the function.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts [hunk 14]
```diff
@@ -169,4 +175,18 @@ export class MultiClusterService {
...
+    }
+    return this.showDeletionMessage;
+  }
+
+  isClusterAdded(isClusterAddedFlag?: boolean) {
+    if (isClusterAddedFlag !== undefined) {
+      this.isClusterAddedFlag = isClusterAddedFlag;
+    }
+    return this.isClusterAddedFlag;
+  }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts:191]: same here

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts [hunk 15]
```diff
@@ -200,36 +206,71 @@ export class PrometheusService {
...
+                  requests.push(request);
+                } else if (ClusterUltilizationQueryList.includes(queryName)) {
+                  const request = this.getMultiClusterQueryRangeData({
+                    params: encodeURIComponent(query),
+                    start: clusterUtilizationChartSelectedTime['start'],
+                    end: clusterUtilizationChartSelectedTime['end'],
+                    step: clusterUtilizationChartSelectedTime['step']
+                  });
+                  requests.push(request);
+                }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:273]: if at some point we introduce more charts apart from PoolUtilization or ClusterUtlization, then we need to grow this if else again. For what seems like a duplicated code, this could have been implemented bit more carefully and generalized way

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html [hunk 16]
```diff
@@ -10,6 +10,12 @@
           [formGroup]="deletionForm"
           novalidate>
       <div class="modal-body">
+        <cd-alert-panel *ngIf="showInfoMessage"
+                        type="info"
+                        spacingClass="mb-3"
+                        i18n>
+          <p>{{ infoMessage }}</p>
+        </cd-alert-panel>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html:18]: why do we need to variables here?

if there is an `infoMessage` then we can show this alertpanel, otherwise we don't need to show it.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.ts [hunk 17]
```diff
@@ -25,6 +25,8 @@ export class CriticalConfirmationModalComponent implements OnInit {
   itemDescription: 'entry';
   itemNames: string[];
   actionDescription = 'delete';
+  showInfoMessage = false;
+  infoMessage: string;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.ts:29]: one would be enough and I prefer `infoMessage`
```suggestion
  infoMessage: string;
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html (hunk 16 above)
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/components/critical-confirmation-modal/critical-confirmation-modal.component.html:18]:
```suggestion
        <cd-alert-panel *ngIf="infoMessage"
                        type="info"
                        spacingClass="mb-3"
                        i18n>
          <p>{{ infoMessage }}</p>
        </cd-alert-panel>
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts [hunk 18]
```diff
@@ -131,26 +104,8 @@ export class ApiInterceptorService implements HttpInterceptor {
...
-                  });
-                } else {
-                  this.router.navigateByUrl('/', { skipLocationChange: true }).then(() => {
-                    this.router.navigate([currentRoute]);
-                  });
-                }
-              } else {
-                this.authStorageService.remove();
-                this.router.navigate(['/login']);
-              }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts:153]: what was the reason this was initially added? and did you verify if that behaviour is working as intended?

File: src/pybind/mgr/dashboard/frontend/src/styles.scss [hunk 19]
```diff
@@ -10,7 +10,7 @@ $font-family-icon: 'ForkAwesome';
 
 $badge-font-size: 1rem;
 $form-feedback-font-size: 100%;
-$popover-max-width: 350px;
+$popover-max-width: 700px;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/styles.scss:13]: we shouldn't do this. If this is something that is needed only for multi-cluster the change it there only unless we know the impact of this change and clearly say that this won't break anything hidden that we don't know of yet

**Review by nizamial09 — COMMENTED** [review]
1. What happens if we never receive the expected prometheus data? Do we show the `Loading` spinner endlessly or at some point we fail and say something's wrong?

something's I've observed.
somehow the url became null when I connected to second cluster.
Screenshot from 2024-03-20 10-56-14
![Screenshot from 2024-03-20 10-51-48]
Screenshot from 2024-03-20 10-47-51
and the config was
```
  "config": {
    "ffcda920-e614-11ee-a7f7-525400869e52": [
      {
        "name": "ffcda920-e614-11ee-a7f7-525400869e52",
        "url": null,
        "cluster_alias": "local-cluster",
        "user": "admin"
      }
    ],
    ....
```

was this something observed before?

File: src/pybind/mgr/dashboard/controllers/auth.py [hunk 20]
```diff
@@ -109,6 +65,49 @@ def create(self, username, password):
...
+                                {
+                                    "name": fsid,
+                                    "url": origin,
+                                    "cluster_alias": "local-cluster",
+                                    "user": username
+                                }
+                            ]
+                        }
+                    }
+                Settings.MULTICLUSTER_CONFIG = multicluster_config
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/controllers/auth.py:110]: This is to check that if a user login with a different user, then we don't store that user's info in the multi-cluster config.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts (hunk 10 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts:223]: When we add a target in the prometheus federation, it takes some time to get the metrics of the target cluster, so to not show incorrect data to the user, this delay is added here.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-area-chart/dashboard-area-chart.component.html (hunk 11 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-area-chart/dashboard-area-chart.component.html:27]: we donot want the chart class to be applied in case of multi-cluster because it restricts the height of the chart

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-pie/dashboard-pie.component.ts (hunk 12 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard-pie/dashboard-pie.component.ts:161]: prometheus query results send the data as string

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts (hunk 13 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts:180]: We need this message for inter communication between the overview component and the cluster list component. If a user disconnects a cluster from the cluster-list page, we should get a flag in the overview page to show a message that disconnected cluster's data shall be removed after ~ 5 minutes

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts (hunk 14 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/multi-cluster.service.ts:191]: same reason as above

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts (hunk 15 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:273]: I can try to generalize it, but since this is working well, i'l not do it in this PR

File: src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts (hunk 18 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/services/api-interceptor.service.ts:153]: This was added to avoid the hub cluster's logout if the token of the remote cluster expires. I have removed it for now. I will add a modified fix in this PR - https://github.com/ceph/ceph/pull/56190

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html [hunk 21]
```diff
@@ -29,7 +29,15 @@
...
-            <cd-helper>Enter the Dashboard API URL. You can retrieve it from the CLI with: <b>ceph mgr services</b></cd-helper>
+            <cd-helper>
+              <cd-alert-panel type="info"
+                              spacingClass="mb-3"
+                              size="normal"
+                              showTitle="false">
+              <p>Enter the Dashboard API URL. You can retrieve it from the CLI with:</p>
+                <cd-code-block [codes]="[clusterApiUrlCmd]"></cd-code-block>
+              </cd-alert-panel>
+            </cd-helper>
```
[the remark this answers is not in the document]
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html]: Updated the PR

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html (hunk 8 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster-form/multi-cluster-form.component.html:145]: Yes, we'l do it in a separate PR

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts (hunk 9 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/multi-cluster/multi-cluster.component.ts]: Updated

**Comment by aaSharma14** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts (hunk 15 above)
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:273]: Why are we doing this? why do we need to check in service that its a pool query or cluster utilization one?

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts [hunk 22]
```diff
@@ -200,36 +206,71 @@ export class PrometheusService {
...
+            if (selectedQueries === 'clusterUtilization') {
+              queries = ClusterUltilizationQueries;
+            }
+            if (selectedQueries === 'poolUtilization') {
+              queries = PoolUltilizationQueries;
+            }
+            if (selectedQueries === 'allQueries') {
+              queries = allQueries;
+            }
+          }
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:250]: same as below

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts [hunk 23]
```diff
@@ -200,36 +206,71 @@ export class PrometheusService {
     return this.http.get<any>(`${this.baseURL}/data`, { params });
   }
 
-  getMultiClusterQueriesData(selectedTime: any, queries: any, queriesResults: any) {
+  getMultiClusterQueriesData(
+    selectedTime: any,
+    poolsChartSelectedTime: any,
+    clusterUtilizationChartSelectedTime: any,
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts:212]: You need to different time values you need to mount them as different objects and it will work.

File: src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts (hunk 15 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts]: Updated the PR

Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/shared/api/prometheus.service.ts]: Updated the PR

File: src/cephadm/cephadm.py
Comment (nizamial09) [inline src/cephadm/cephadm.py:1]: this file shouldn't be here.

**Review by nizamial09 — CHANGES_REQUESTED** [review]

File: src/cephadm/cephadm.py
Comment (aaSharma14) [reply src/cephadm/cephadm.py:1]: Removed @nizamial09

**Comment by aaSharma14** [comment]
jenkins test api

**Review by nizamial09 — APPROVED** [review]
